c-dependencies/js-compute-runtime/builtins/secret-store.cpp

```diff
@@ -0,0 +1,197 @@
+  }
+
+  JS::SetReservedSlot(entry, Slots::Handle, JS::Int32Value(handle));
+
+  return entry;
+}
+
+bool SecretStoreEntry::init_class(JSContext *cx, JS::HandleObject global) {
+  return BuiltinImpl<SecretStoreEntry>::init_class_impl(cx, global);
+}
+
+fastly_secret_store_handle_t SecretStore::secret_store_handle(JSObject *obj) {
+  JS::Value val = JS::GetReservedSlot(obj, SecretStore::Slots::Handle);
+  return static_cast<fastly_secret_store_handle_t>(val.toInt32());
+}
+
+bool SecretStore::get(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(1)
+
+  xqd_world_string_t key_str;
+  JS::UniqueChars key = encode(cx, args[0], &key_str.len);
+  if (!key) {
+    return false;
+  }
+  // If the converted string has a length of 0 then we throw an Error
+  // because keys have to be at-least 1 character.
+  if (!key || key_str.len == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_SECRET_STORE_KEY_EMPTY);
+    return false;
+  }
+  key_str.ptr = key.get();
+
+  // key has to be less than 256
+  if (key_str.len > 255) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_SECRET_STORE_KEY_TOO_LONG);
+    return false;
+  }
+
+  std::string_view keyView(key_str.ptr, key_str.len);
+
+  // key must contain only letters, numbers, dashes (-), underscores (_), and periods (.).
+  auto is_valid_key = std::all_of(keyView.begin(), keyView.end(), [&](auto character) {
+    return std::isalnum(character) || character == '_' || character == '-' || character == '.';
+  });
+
+  if (!is_valid_key) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_SECRET_STORE_KEY_CONTAINS_INVALID_CHARACTER);
+    return false;
+  }
+
+  fastly_option_secret_handle_t ret;
+  fastly_error_t err;
+  // Ensure that we throw an exception for all unexpected host errors.
+  if (!xqd_fastly_secret_store_get(SecretStore::secret_store_handle(self), &key_str, &ret, &err)) {
+    HANDLE_ERROR(cx, err);
+    return false;
+  }
+
+  // None indicates the key wasn't found, so we return null.
+  if (!ret.is_some) {
+    args.rval().setNull();
+    return true;
+  }
+
+  // When no entry is found, we are going to resolve the Promise with `null`.
+  if (!ret.is_some) {
+    args.rval().setNull();
+  } else {
+    JS::RootedObject entry(cx, SecretStoreEntry::create(cx, ret.val));
+    if (!entry) {
+      return false;
+    }
+    args.rval().setObject(*entry);
+  }
+
+  return true;
+}
+
+const JSFunctionSpec SecretStore::methods[] = {JS_FN("get", get, 1, JSPROP_ENUMERATE), JS_FS_END};
+
+const JSPropertySpec SecretStore::properties[] = {JS_PS_END};
+
+bool SecretStore::constructor(JSContext *cx, unsigned argc, JS::Value *vp) {
+  REQUEST_HANDLER_ONLY("The SecretStore builtin");
+  CTOR_HEADER("SecretStore", 1);
+
+  xqd_world_string_t name_str;
+  JS::UniqueChars name_chars = encode(cx, args[0], &name_str.len);
+  if (!name_chars) {
+    return false;
+  }
+  name_str.ptr = name_chars.get();
```

Comment:
Same comment here, I'd move the declaration of `name_str` down to where it's used on line 175 to make it clear that `name_chars` is the unique owner of the memory.